File: public/api/chat_threads.php
<?php
declare(strict_types=1);

header('Content-Type: application/json; charset=utf-8');

require_once __DIR__ . '/../includes/bootstrap.php';
app_require_login();

$pdo = ventas_pdo();
$username = trim((string) ($_SESSION['usuario'] ?? ''));
if ($username === '') {
    http_response_code(401);
    echo json_encode(['ok' => false, 'error' => 'Sesión inválida'], JSON_UNESCAPED_UNICODE);
    exit;
}

function json_body(): array {
    $raw = file_get_contents('php://input') ?: '';
    $data = json_decode($raw, true);
    return is_array($data) ? $data : [];
}

try {
    $method = strtoupper((string) ($_SERVER['REQUEST_METHOD'] ?? 'GET'));

    if ($method === 'GET') {
        $clientId = isset($_GET['thread']) ? trim((string) $_GET['thread']) : '';
        $q = isset($_GET['q']) ? trim((string) $_GET['q']) : '';

        if ($clientId !== '') {
            $st = $pdo->prepare('SELECT id, client_thread_id, title, created_at, updated_at FROM app_chat_threads WHERE username = :u AND client_thread_id = :cid LIMIT 1');
            $st->execute([':u' => $username, ':cid' => $clientId]);
            $t = $st->fetch(PDO::FETCH_ASSOC) ?: null;
            if (!$t) {
                echo json_encode(['ok' => true, 'thread' => null], JSON_UNESCAPED_UNICODE | JSON_INVALID_UTF8_SUBSTITUTE);
                exit;
            }

            $stM = $pdo->prepare('SELECT role, content, created_at FROM app_chat_messages WHERE thread_id = :tid ORDER BY id ASC');
            $stM->execute([':tid' => (int) $t['id']]);
            $msgs = [];
            while ($r = $stM->fetch(PDO::FETCH_ASSOC)) {
                $msgs[] = [
                    'role' => (string) ($r['role'] ?? ''),
                    'content' => (string) ($r['content'] ?? ''),
                    'createdAt' => (string) ($r['created_at'] ?? ''),
                ];
            }

            echo json_encode([
                'ok' => true,
                'thread' => [
                    'id' => (string) ($t['client_thread_id'] ?? ''),
                    'title' => (string) ($t['title'] ?? ''),
                    'createdAt' => (string) ($t['created_at'] ?? ''),
                    'updatedAt' => (string) ($t['updated_at'] ?? ''),
                    'messages' => $msgs,
                ],
            ], JSON_UNESCAPED_UNICODE | JSON_INVALID_UTF8_SUBSTITUTE);
            exit;
        }

        // Listado
        $sql = 'SELECT t.client_thread_id, t.title, t.updated_at,
                       (SELECT COUNT(*) FROM app_chat_messages m WHERE m.thread_id = t.id) AS n
                FROM app_chat_threads t
                WHERE t.username = :u';
        $params = [':u' => $username];
        if ($q !== '') {
            $sql .= ' AND (t.title LIKE :q OR EXISTS (
                        SELECT 1 FROM app_chat_messages m2
                        WHERE m2.thread_id = t.id AND m2.content LIKE :q
                    ))';
            $params[':q'] = '%' . $q . '%';
        }
        $sql .= ' ORDER BY t.updated_at DESC LIMIT 60';

        $st = $pdo->prepare($sql);
        $st->execute($params);
        $threads = [];
        while ($r = $st->fetch(PDO::FETCH_ASSOC)) {
            $threads[] = [
                'id' => (string) ($r['client_thread_id'] ?? ''),
                'title' => (string) ($r['title'] ?? ''),
                'updatedAt' => (string) ($r['updated_at'] ?? ''),
                'n' => (int) ($r['n'] ?? 0),
            ];
        }

        echo json_encode(['ok' => true, 'threads' => $threads], JSON_UNESCAPED_UNICODE | JSON_INVALID_UTF8_SUBSTITUTE);
        exit;
    }

    if ($method === 'POST') {
        $body = json_body();
        $clientId = isset($body['id']) ? trim((string) $body['id']) : '';
        $title = isset($body['title']) ? trim((string) $body['title']) : 'Nuevo chat';
        $messages = isset($body['messages']) && is_array($body['messages']) ? $body['messages'] : [];

        if ($clientId === '') {
            http_response_code(400);
            echo json_encode(['ok' => false, 'error' => 'Falta id'], JSON_UNESCAPED_UNICODE);
            exit;
        }
        if ($title === '') $title = 'Nuevo chat';
        if (strlen($title) > 220) $title = substr($title, 0, 220);

        $pdo->beginTransaction();

        $st = $pdo->prepare('SELECT id FROM app_chat_threads WHERE username = :u AND client_thread_id = :cid LIMIT 1');
        $st->execute([':u' => $username, ':cid' => $clientId]);
        $row = $st->fetch(PDO::FETCH_ASSOC) ?: null;

        if ($row) {
            $threadId = (int) $row['id'];
            $stU = $pdo->prepare('UPDATE app_chat_threads SET title = :t, updated_at = NOW() WHERE id = :id');
            $stU->execute([':t' => $title, ':id' => $threadId]);
            $pdo->prepare('DELETE FROM app_chat_messages WHERE thread_id = :id')->execute([':id' => $threadId]);
        } else {
            $stI = $pdo->prepare('INSERT INTO app_chat_threads (username, client_thread_id, title) VALUES (:u, :cid, :t)');
            $stI->execute([':u' => $username, ':cid' => $clientId, ':t' => $title]);
            $threadId = (int) $pdo->lastInsertId();
        }

        $stM = $pdo->prepare('INSERT INTO app_chat_messages (thread_id, role, content) VALUES (:tid, :role, :content)');
        $n = 0;
        foreach ($messages as $m) {
            if (!is_array($m)) continue;
            $role = (string) ($m['role'] ?? '');
            if ($role !== 'user' && $role !== 'assistant') continue;
            $content = (string) ($m['content'] ?? '');
            if ($content === '') continue;
            // MEDIUMTEXT en BD; evitamos truncar respuestas largas del asistente (listas + SQL).
            if (strlen($content) > 524288) {
                $content = substr($content, 0, 524288);
            }
            $stM->execute([':tid' => $threadId, ':role' => $role, ':content' => $content]);
            $n++;
            if ($n >= 500) {
                break;
            }
        }

        $pdo->commit();
        echo json_encode(['ok' => true], JSON_UNESCAPED_UNICODE);
        exit;
    }

    if ($method === 'DELETE') {
        $body = json_body();
        $clientId = isset($body['id']) ? trim((string) $body['id']) : '';
        if ($clientId === '') {
            http_response_code(400);
            echo json_encode(['ok' => false, 'error' => 'Falta id'], JSON_UNESCAPED_UNICODE);
            exit;
        }
        $st = $pdo->prepare('DELETE FROM app_chat_threads WHERE username = :u AND client_thread_id = :cid');
        $st->execute([':u' => $username, ':cid' => $clientId]);
        echo json_encode(['ok' => true], JSON_UNESCAPED_UNICODE);
        exit;
    }

    http_response_code(405);
    echo json_encode(['ok' => false, 'error' => 'Método no permitido'], JSON_UNESCAPED_UNICODE);
} catch (Throwable $e) {
    if ($pdo->inTransaction()) {
        try { $pdo->rollBack(); } catch (Throwable) {}
    }
    http_response_code(500);
    echo json_encode(['ok' => false, 'error' => $e->getMessage()], JSON_UNESCAPED_UNICODE | JSON_INVALID_UTF8_SUBSTITUTE);
}

File: backend_python/routes/api/chat_threads.py
from flask import Blueprint, request, session, jsonify
from services.db import get_connection

bp = Blueprint('api_chat_threads', __name__)


def _require_login():
    if not session.get('active'):
        return jsonify({'ok': False, 'error': 'No autenticado'}), 401
    username = str(session.get('usuario') or '').strip()
    if not username:
        return jsonify({'ok': False, 'error': 'Sesión inválida'}), 401
    return None, username


@bp.route('/api/chat_threads.php', methods=['GET', 'POST', 'DELETE'])
def chat_threads():
    err_resp = _require_login()
    if err_resp[0] is not None:
        return err_resp
    _, username = err_resp

    conn = get_connection()
    method = request.method

    if method == 'GET':
        return _get(conn, username)
    if method == 'POST':
        return _post(conn, username)
    if method == 'DELETE':
        return _delete(conn, username)

    return jsonify({'ok': False, 'error': 'Método no permitido'}), 405


def _get(conn, username):
    client_id = request.args.get('thread', '').strip()
    q = request.args.get('q', '').strip()

    if client_id:
        with conn.cursor() as cur:
            cur.execute(
                'SELECT id, client_thread_id, title, created_at, updated_at FROM app_chat_threads WHERE username = %s AND client_thread_id = %s LIMIT 1',
                (username, client_id)
            )
            t = cur.fetchone()
        if not t:
            return jsonify({'ok': True, 'thread': None})

        with conn.cursor() as cur:
            cur.execute(
                'SELECT role, content, created_at FROM app_chat_messages WHERE thread_id = %s ORDER BY id ASC',
                (int(t['id']),)
            )
            msgs = [{'role': str(r['role'] or ''), 'content': str(r['content'] or ''), 'createdAt': str(r['created_at'] or '')}
                    for r in cur.fetchall()]

        return jsonify({'ok': True, 'thread': {
            'id': str(t['client_thread_id'] or ''),
            'title': str(t['title'] or ''),
            'createdAt': str(t['created_at'] or ''),
            'updatedAt': str(t['updated_at'] or ''),
            'messages': msgs,
        }})

    sql = ('SELECT t.client_thread_id, t.title, t.updated_at,'
           ' (SELECT COUNT(*) FROM app_chat_messages m WHERE m.thread_id = t.id) AS n'
           ' FROM app_chat_threads t WHERE t.username = %s')
    params = [username]
    if q:
        sql += (' AND (t.title LIKE %s OR EXISTS ('
                ' SELECT 1 FROM app_chat_messages m2 WHERE m2.thread_id = t.id AND m2.content LIKE %s'
                '))')
        params += [f'%{q}%', f'%{q}%']
    sql += ' ORDER BY t.updated_at DESC LIMIT 60'

    with conn.cursor() as cur:
        cur.execute(sql, params)
        threads = [{'id': str(r['client_thread_id'] or ''), 'title': str(r['title'] or ''),
                    'updatedAt': str(r['updated_at'] or ''), 'n': int(r['n'] or 0)}
                   for r in cur.fetchall()]

    return jsonify({'ok': True, 'threads': threads})


def _post(conn, username):
    body = request.get_json(silent=True) or {}
    client_id = str(body.get('id') or '').strip()
    title = str(body.get('title') or '').strip() or 'Nuevo chat'
    messages = body.get('messages') if isinstance(body.get('messages'), list) else []

    if not client_id:
        return jsonify({'ok': False, 'error': 'Falta id'}), 400
    if len(title) > 220:
        title = title[:220]

    conn.begin()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'SELECT id FROM app_chat_threads WHERE username = %s AND client_thread_id = %s LIMIT 1',
                (username, client_id)
            )
            row = cur.fetchone()

        if row:
            thread_id = int(row['id'])
            with conn.cursor() as cur:
                cur.execute('UPDATE app_chat_threads SET title = %s, updated_at = NOW() WHERE id = %s', (title, thread_id))
                cur.execute('DELETE FROM app_chat_messages WHERE thread_id = %s', (thread_id,))
        else:
            with conn.cursor() as cur:
                cur.execute(
                    'INSERT INTO app_chat_threads (username, client_thread_id, title) VALUES (%s, %s, %s)',
                    (username, client_id, title)
                )
                thread_id = cur.lastrowid

        n = 0
        with conn.cursor() as cur:
            for m in messages:
                if not isinstance(m, dict):
                    continue
                role = str(m.get('role') or '')
                if role not in ('user', 'assistant'):
                    continue
                content = str(m.get('content') or '')
                if not content:
                    continue
                if len(content) > 524288:
                    content = content[:524288]
                cur.execute(
                    'INSERT INTO app_chat_messages (thread_id, role, content) VALUES (%s, %s, %s)',
                    (thread_id, role, content)
                )
                n += 1
                if n >= 500:
                    break

        conn.commit()
        return jsonify({'ok': True})
    except Exception as e:
        try:
            conn.rollback()
        except Exception:
            pass
        return jsonify({'ok': False, 'error': str(e)}), 500


def _delete(conn, username):
    body = request.get_json(silent=True) or {}
    purge_all = bool(body.get('purge_all'))

    if purge_all:
        with conn.cursor() as cur:
            cur.execute('DELETE FROM app_chat_threads WHERE username = %s', (username,))
        return jsonify({'ok': True, 'purge_all': True})

    client_id = str(body.get('id') or '').strip()
    if not client_id:
        return jsonify({'ok': False, 'error': 'Falta id'}), 400

    with conn.cursor() as cur:
        cur.execute('DELETE FROM app_chat_threads WHERE username = %s AND client_thread_id = %s', (username, client_id))
    return jsonify({'ok': True})
